docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+    >>>     return set(sum((node.inputs + node.outputs for node in apply_nodes), []))
+    >>> 
+    >>>     
+    >>> inputs = tensor.matrix('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(10, 754))
+    >>> dims = [754, 100, 100, 10]
+    >>> activations = [tensor.tanh, tensor.tanh, tensor.nnet.softmax]
+    >>> weight_initializers = [
+    >>>     lambda shape: numpy.random.uniform(size=shape)] * len(activations)
+    >>> bias_initializers = weight_initializers
+    >>> stone = MultiLayer(
+    >>>     dims, activations, weight_initializers, bias_initializers)
+    >>> outputs_multilayer = stone.apply(inputs)   
+    >>> layer0_outputs, = [v for v in get_all_variables([outputs_multilayer]) 
+    >>>                    if v.name == 'layer_0_activations']
+    >>> theano.printing.debugprint(layer0_outputs)
+    Elemwise{tanh,no_inplace} [@A] 'layer_0_activations'   
+     |Elemwise{add,no_inplace} [@B] ''   
+       |dot [@C] ''   
+       | |x [@D]
+       | |<TensorType(float64, matrix)> [@E]
+       |DimShuffle{x,0} [@F] ''   
+         |<TensorType(float64, vector)> [@G]
+
+
+Great, it works! But something tells D that when the library has
+grown up and consists of numerous components, it will be very hard to
+keep track of various name mangling schemes, such as the one used in the
+example above. A more systematic way of annotating the graph is
+desirable.
+
+Here is an idea: if the activations functions were stones as well, and
+if input and outputs variables of all stones were respectively marked,
+such annotation would be sufficient for D's purposes. For instance,
+the ``Tanh`` stone could look as follows:
+
+    >>> class Tanh(Stone):
+    >>>     def apply(self, x):
+    >>>         x = x.copy()
+    >>>         # The copying is necessary because the variable x might already be 
+    >>>         # an output of another stone. Potentially, one could annotate 
+    >>>         # variable as belonging to several stones, but the way we go here
+    >>>         # seems much simpler. Under the hood `x.copy()` is an element-wise copy
+    >>>         # of x, and all such variables should be quickly removed by the optimizer.
+    >>>         x.tag.stone = self
+    >>>         x.tag.role = 'input'
+    >>>         # We use the `.tag` of Theano variables to add various annotation,
+    >>>         # such the stone which created the variable and its "role", that is
+    >>>         # whether it was an input of the stone or an output.
+    >>>         y = tensor.tanh(x).copy()
+    >>>         y.tag.stone = self
+    >>>         y.tag.role = 'output'
+    >>>         return y    
+
+It seems that the process of annotating could be largely automated.
+Arguably, the most convenient way to proceed is to use decorators.
+Decorators belong to the realm of somewhat less-known Python features,
+but nevertheless, they are very simple:
+
+    >>> def application(wrapped_method):
```

Comment:
I agree with Dima: this is one of the most obscure parts of Blocks (Python, actually), where people get puzzled and move to something simpler to understand. It's very important to comment it clearly and in depth IMO.